#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/RectangleArena.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/robots/k-team/khepera1/protos/Khepera1.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/Floor.proto"

WorldInfo {
  info [
    "A Khepera robot equipped with a gripper"
  ]
  title "Khepera Gripper"
  CFM 0.001
  ERP 0.4
  basicTimeStep 10
  contactProperties [
    ContactProperties {
      material1 "gripper"
      material2 "stick"
      coulombFriction [
        10
      ]
    }
    ContactProperties {
      material1 "khepera"
      coulombFriction [
        0.1
      ]
    }
  ]
}
Viewpoint {
  orientation -0.4793872236263668 0.47634364328753337 0.7370784377007396 1.877316854360154
  position 0.016977151093250382 -1.1039934713358044 2.697130471405026
}
TexturedBackground {
}
TexturedBackgroundLight {
}
RectangleArena {
  translation 0 0 -2.26951e-18
  rotation 0 1 0 4.692820414042842e-06
  floorSize 1.5 1.5
}
DEF OverheadCamera Robot {
  translation 0 0 2
  rotation 0 -1 0 -1.57079
  children [
    Camera {
      rotation 1 0 0 -1.5708
      name "overhead_camera"
      fieldOfView 0.85
      width 1024
      height 1024
    }
    Emitter {
      name "emitter1"
      channel 1
    }
    Emitter {
      name "emitter2"
      channel 1
    }
    Receiver {
      name "receiver1"
      channel 11
    }
    Receiver {
      name "receiver2"
      channel 12
    }
  ]
  name "overhead_camera"
  controller "supervisor_debug_messages"
  supervisor TRUE
}
Khepera1 {
  translation -0.4818736099196599 -0.599553965075449 0.00010246378181508198
  rotation 0.035070953861541264 -0.0643929048891237 0.997308168017881 0.999719917730578
  name "khepera"
  controller "debug_messages"
  extensionSlot [
    DEF LIFT_HINGE_JOINT HingeJoint {
      jointParameters HingeJointParameters {
        position -1.3999999866264539
        axis 0 1 0
        anchor -0.01 0 0.022
      }
      device [
        RotationalMotor {
          name "motor"
        }
      ]
      endPoint Solid {
        translation -0.01 0 0.022
        rotation 0.7058148422121028 -0.7058148422121053 -0.060421991246612516 3.0208949905607376
        children [
          DEF RIGHT_GRIP SliderJoint {
            jointParameters JointParameters {
              position 0.004999999999891248
              axis 1 0 0
            }
            device [
              LinearMotor {
                name "right grip"
                maxVelocity 0.1
                maxPosition 0.029000000000765386
                maxForce 100
              }
            ]
            endPoint Solid {
              translation 0.004999999999891248 0 -0.048
              rotation 0 1 0 1.5708
              children [
                Shape {
                  appearance PBRAppearance {
                    baseColor 0.8 0.8 0.8
                    roughness 1
                    metalness 0
                  }
                  geometry IndexedFaceSet {
                    coord Coordinate {
                      point [
                        0 -0.004 0.004
                        0.03 -0.004 0.002
                        0.03 -0.004 0.0001
                        0 -0.004 0.0001
                        0 -0.004 0.004
                        0 0.004 0.004
                        0.03 0.004 0.002
                        0.03 0.004 0.0001
                        0 0.004 0.0001
                        0 0.004 0.004
                      ]
                    }
                    coordIndex [
                      4, 3, 2, 1, 0, -1, 5, 6, 7, 8, 9, -1, 0, 1, 6, 5, -1, 1, 2, 7, 6, -1, 2, 3, 8, 7, -1, 3, 4, 9, 8, -1
                    ]
                  }
                }
                DistanceSensor {
                  translation 0.015 0 0
                  rotation 0 1 0 1.5708
                  name "ds"
                  lookupTable [
                    0 0 0
                    0.05 1000 0
                  ]
                  type "infra-red"
                }
              ]
              name "right grip"
              contactMaterial "gripper"
              boundingObject Pose {
                translation 0.015 0 0.003
                children [
                  Box {
                    size 0.028 0.007 0.005
                  }
                ]
              }
              physics Physics {
                density 100
                centerOfMass [
                  0.01 0 0
                ]
              }
            }
          }
          DEF LEFT_GRIP SliderJoint {
            jointParameters JointParameters {
              position 0.005000000000108556
              axis -1 0 0
            }
            device [
              LinearMotor {
                name "left grip"
                maxVelocity 0.1
                maxPosition 0.02900000000091496
                maxForce 100
              }
            ]
            endPoint Solid {
              translation -0.005000000000108556 0 -0.048
              rotation 0 1 0 -1.5708
              children [
                Shape {
                  appearance PBRAppearance {
                    baseColor 0.8 0.8 0.8
                    roughness 1
                    metalness 0
                  }
                  geometry IndexedFaceSet {
                    coord Coordinate {
                      point [
                        0 -0.004 0.0001
                        -0.03 -0.004 0.0001
                        -0.03 -0.004 0.002
                        0 -0.004 0.004
                        0 -0.004 0.0001
                        0 0.004 0.0001
                        -0.03 0.004 0.0001
                        -0.03 0.004 0.002
                        0 0.004 0.004
                        0 0.004 0.0001
                      ]
                    }
                    coordIndex [
                      4, 3, 2, 1, 0, -1, 5, 6, 7, 8, 9, -1, 0, 1, 6, 5, -1, 1, 2, 7, 6, -1, 2, 3, 8, 7, -1, 3, 4, 9, 8, -1
                    ]
                  }
                }
              ]
              name "left grip"
              contactMaterial "gripper"
              boundingObject Pose {
                translation -0.015 0 0.003
                children [
                  Box {
                    size 0.028 0.007 0.005
                  }
                ]
              }
              physics Physics {
                density 100
                centerOfMass [
                  -0.01 0 0
                ]
              }
            }
          }
          Shape {
            appearance PBRAppearance {
              baseColor 0.8 0.8 0.8
              roughness 1
              metalness 0
            }
            geometry IndexedFaceSet {
              coord Coordinate {
                point [
                  0.035 -0.004 0.006
                  0.035 -0.004 -0.048
                  -0.035 -0.004 -0.048
                  -0.035 -0.004 0.006
                  -0.025 -0.004 0.006
                  -0.025 -0.004 -0.03
                  -0.016 -0.004 -0.04
                  0.016 -0.004 -0.04
                  0.025 -0.004 -0.03
                  0.025 -0.004 0.006
                  0.035 -0.004 0.006
                  0.035 0.004 0.006
                  0.035 0.004 -0.048
                  -0.035 0.004 -0.048
                  -0.035 0.004 0.006
                  -0.025 0.004 0.006
                  -0.025 0.004 -0.03
                  -0.016 0.004 -0.04
                  0.016 0.004 -0.04
                  0.025 0.004 -0.03
                  0.025 0.004 0.006
                  0.035 0.004 0.006
                ]
              }
              coordIndex [
                10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, -1, 0, 1, 12, 11, -1, 1, 2, 13, 12, -1, 2, 3, 14, 13, -1, 3, 4, 15, 14, -1, 4, 5, 16, 15, -1, 5, 6, 17, 16, -1, 6, 7, 18, 17, -1, 7, 8, 19, 18, -1, 8, 9, 20, 19, -1, 9, 10, 21, 20, -1
              ]
            }
          }
        ]
        boundingObject Pose {
          translation 0 0 -0.021
          children [
            Box {
              size 0.068 0.008 0.052
            }
          ]
        }
        physics Physics {
          density 100
          centerOfMass [
            0 0 -0.02
          ]
        }
      }
    }
    DEF GRIPPER_BASE Pose {
      translation -0.01 0 0.023
      rotation 0.5773502691896258 -0.5773502691896258 -0.5773502691896258 2.094395
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.8 0.8 0.8
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.05 0.01 0.01
          }
        }
      ]
    }
    Camera {
      translation -0.01 0 0.035
      rotation 1 0 0 0
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.70632 0.70632 0.70632
            roughness 1
            metalness 0
          }
          geometry DEF CAMERA Box {
            size 0.03 0.02 0.03
          }
        }
      ]
      boundingObject DEF CAMERA Box {
        size 0.03 0.02 0.03
      }
      physics Physics {
        density -1
        mass 0.02
      }
    }
    Emitter {
      rotation -0.05037859281586198 0.05037871387365086 0.99745876234275 4.70984457625572
      channel 11
    }
    Receiver {
      rotation -0.06801484702383666 0.06801500910925684 -0.9953632196942941 1.5754439950300838
      channel 1
    }
  ]
}
Khepera1 {
  translation 0.6137169900319405 0.5590872877542963 0.00010246378181487555
  rotation 0.03514131471619386 0.013518262612683083 0.9992909208913848 -2.4075863378353093
  name "khepera(1)"
  controller "debug_messages2"
  extensionSlot [
    DEF LIFT_HINGE_JOINT HingeJoint {
      jointParameters HingeJointParameters {
        position -1.3999999866263209
        axis 0 1 0
        anchor -0.01 0 0.022
      }
      device [
        RotationalMotor {
          name "motor"
        }
      ]
      endPoint Solid {
        translation -0.01 0 0.022
        rotation 0.7058148422121008 -0.7058148422121032 -0.060421991246659555 3.0208949905606444
        children [
          DEF RIGHT_GRIP SliderJoint {
            jointParameters JointParameters {
              position 0.005000000002457641
              axis 1 0 0
            }
            device [
              LinearMotor {
                name "right grip"
                maxVelocity 0.1
                maxPosition 0.029000000000765386
                maxForce 100
              }
            ]
            endPoint Solid {
              translation 0.005000000002457641 0 -0.048
              rotation 0 1 0 1.5708
              children [
                Shape {
                  appearance PBRAppearance {
                    baseColor 0.8 0.8 0.8
                    roughness 1
                    metalness 0
                  }
                  geometry IndexedFaceSet {
                    coord Coordinate {
                      point [
                        0 -0.004 0.004
                        0.03 -0.004 0.002
                        0.03 -0.004 0.0001
                        0 -0.004 0.0001
                        0 -0.004 0.004
                        0 0.004 0.004
                        0.03 0.004 0.002
                        0.03 0.004 0.0001
                        0 0.004 0.0001
                        0 0.004 0.004
                      ]
                    }
                    coordIndex [
                      4, 3, 2, 1, 0, -1, 5, 6, 7, 8, 9, -1, 0, 1, 6, 5, -1, 1, 2, 7, 6, -1, 2, 3, 8, 7, -1, 3, 4, 9, 8, -1
                    ]
                  }
                }
                DistanceSensor {
                  translation 0.015 0 0
                  rotation 0 1 0 1.5708
                  name "ds"
                  lookupTable [
                    0 0 0
                    0.05 1000 0
                  ]
                  type "infra-red"
                }
              ]
              name "right grip"
              contactMaterial "gripper"
              boundingObject Pose {
                translation 0.015 0 0.003
                children [
                  Box {
                    size 0.028 0.007 0.005
                  }
                ]
              }
              physics Physics {
                density 100
                centerOfMass [
                  0.01 0 0
                ]
              }
            }
          }
          DEF LEFT_GRIP SliderJoint {
            jointParameters JointParameters {
              position 0.005000000013694695
              axis -1 0 0
            }
            device [
              LinearMotor {
                name "left grip"
                maxVelocity 0.1
                maxPosition 0.02900000000091496
                maxForce 100
              }
            ]
            endPoint Solid {
              translation -0.005000000013694695 0 -0.048
              rotation 0 1 0 -1.5708
              children [
                Shape {
                  appearance PBRAppearance {
                    baseColor 0.8 0.8 0.8
                    roughness 1
                    metalness 0
                  }
                  geometry IndexedFaceSet {
                    coord Coordinate {
                      point [
                        0 -0.004 0.0001
                        -0.03 -0.004 0.0001
                        -0.03 -0.004 0.002
                        0 -0.004 0.004
                        0 -0.004 0.0001
                        0 0.004 0.0001
                        -0.03 0.004 0.0001
                        -0.03 0.004 0.002
                        0 0.004 0.004
                        0 0.004 0.0001
                      ]
                    }
                    coordIndex [
                      4, 3, 2, 1, 0, -1, 5, 6, 7, 8, 9, -1, 0, 1, 6, 5, -1, 1, 2, 7, 6, -1, 2, 3, 8, 7, -1, 3, 4, 9, 8, -1
                    ]
                  }
                }
              ]
              name "left grip"
              contactMaterial "gripper"
              boundingObject Pose {
                translation -0.015 0 0.003
                children [
                  Box {
                    size 0.028 0.007 0.005
                  }
                ]
              }
              physics Physics {
                density 100
                centerOfMass [
                  -0.01 0 0
                ]
              }
            }
          }
          Shape {
            appearance PBRAppearance {
              baseColor 0.8 0.8 0.8
              roughness 1
              metalness 0
            }
            geometry IndexedFaceSet {
              coord Coordinate {
                point [
                  0.035 -0.004 0.006
                  0.035 -0.004 -0.048
                  -0.035 -0.004 -0.048
                  -0.035 -0.004 0.006
                  -0.025 -0.004 0.006
                  -0.025 -0.004 -0.03
                  -0.016 -0.004 -0.04
                  0.016 -0.004 -0.04
                  0.025 -0.004 -0.03
                  0.025 -0.004 0.006
                  0.035 -0.004 0.006
                  0.035 0.004 0.006
                  0.035 0.004 -0.048
                  -0.035 0.004 -0.048
                  -0.035 0.004 0.006
                  -0.025 0.004 0.006
                  -0.025 0.004 -0.03
                  -0.016 0.004 -0.04
                  0.016 0.004 -0.04
                  0.025 0.004 -0.03
                  0.025 0.004 0.006
                  0.035 0.004 0.006
                ]
              }
              coordIndex [
                10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, -1, 0, 1, 12, 11, -1, 1, 2, 13, 12, -1, 2, 3, 14, 13, -1, 3, 4, 15, 14, -1, 4, 5, 16, 15, -1, 5, 6, 17, 16, -1, 6, 7, 18, 17, -1, 7, 8, 19, 18, -1, 8, 9, 20, 19, -1, 9, 10, 21, 20, -1
              ]
            }
          }
        ]
        boundingObject Pose {
          translation 0 0 -0.021
          children [
            Box {
              size 0.068 0.008 0.052
            }
          ]
        }
        physics Physics {
          density 100
          centerOfMass [
            0 0 -0.02
          ]
        }
      }
    }
    DEF GRIPPER_BASE Pose {
      translation -0.01 0 0.023
      rotation 0.5773502691896258 -0.5773502691896258 -0.5773502691896258 2.094395
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.8 0.8 0.8
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.05 0.01 0.01
          }
        }
      ]
    }
    Camera {
      translation -0.01 0 0.035
      rotation 1 0 0 0
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.70632 0.70632 0.70632
            roughness 1
            metalness 0
          }
          geometry DEF CAMERA Box {
            size 0.03 0.02 0.03
          }
        }
      ]
      boundingObject DEF CAMERA Box {
        size 0.03 0.02 0.03
      }
      physics Physics {
        density -1
        mass 0.02
      }
    }
    Emitter {
      rotation -0.05037859281586198 0.05037871387365086 0.99745876234275 4.70984457625572
      channel 12
    }
    Receiver {
      rotation -0.06801484702383666 0.06801500910925684 -0.9953632196942941 1.5754439950300838
      channel 2
    }
  ]
}
Floor {
  translation 0 0 0.0001
  size 1 1
  appearance PBRAppearance {
    baseColorMap ImageTexture {
      url [
        "../textures/squares4x4.png"
      ]
    }
    roughness 0.1
    metalness 0
    textureTransform TextureTransform {
      scale 0.5 0.5
    }
  }
}
